# 1. What is the difference between a function and a method in Python?

### Function

> A function is a block of reusable code that performs specific task and can be called independently.

> Function are defined using the 'def' Keyword.

> Functions can be difined outside of classes and can be called by their name, with or without arguments.

In [5]:
# Example:

def greet(name): # parameter
    return f"Hello, {name}!"

# calling or invokig function
# u can call function as many times
Person1 = greet("Veliston")# argument
Person2 = greet("Rocha")
Person3 = greet("Rodrigues")

print(Person1)
print(Person2)
print(Person3)

Hello, Veliston!
Hello, Rocha!
Hello, Rodrigues!


### Method

> A method is a function that is associated with an object or class. It operates on the data within the class. (object or instance)

>Methods are defined within a class and usually take ' self ' as their parameter to acsess the instance of the class.

> Methods are called on an instance of a class or on the class itself

In [7]:
# Example:

class Greeter:
    def __init__(self, name): # called automaticaly
        self.name = name

    def greet(self): # called maually
        return f"Hello, {self.name}!"

g = Greeter("Veliston")
print(g.greet())  

# greeter is class
# __init__ and greet are methods
# name is attribute


Hello, Veliston!


# 2. Explain the concept of function arguments and parameters in python.**

### Parameters

> Parameters are variables listed in a function's definition that act as placeholders for the values that will be passed to the function when it is called.

> They define what kind of values the function expects to receive.

In [9]:
# example:

def add(a, b):
    return a + b

# a and b are parameters of the ' add ' function

### Arguments

> Arguments are the actual values or expressions passed to a function when it is called. They are assigned to the corresponding parameters in the function definition.

> They provide the actual data for the function to work with.

In [10]:
# example:

result = add(5,3)

# 5 and 3 are arguments passed to the ' add ' function.
# They are assigned to the parameters a and b respectively.

# 3. What are the differnt ways to define and call a function in Python?

### Defining and Calling a Standard Function

> using def keyword to define a function with a specific name and parameters.

> after that calling or invoking the function with arguments.

In [11]:
# Defining
def greet(name):
    return f"Hello, {name}!"

# Calling
print(greet("Veliston Rocha"))

Hello, Veliston Rocha!


### Lambda Functions

> Lambda functions are small, unknown functions defined using the 'lambda' keyword. often used for short fuctions.

> calling with arguments

In [13]:
# defining
add = lambda x, y: x + y

# calling
print(add(5, 3)) 


8


### Functions with Default Arguments

> Functions that have default values for some parameters. If no argument is provided the default value is used.

> u can omit or dont provide argument that have default values.

In [15]:
def greet(name="Person"):
    return f"Hello, {name}!"

print(greet())         # Uses default parameter argument
print(greet("Veliston"))   # Uses provided parameter argument



Hello, Person!
Hello, Veliston!


### Variable-Length Arguments

> These allow you to pass a variable number of arguments to a function.

> *args: Collects positional arguments into a tuple.

> **kwargs: Collects keyword arguments into a dictionary.

In [16]:
# args
def summarize(*args):
    return sum(args)

# kwargs
def describe(**kwargs):
    return kwargs

print(summarize(1, 2, 3, 4)) 
print(describe(name="Veliston", age=21))

10
{'name': 'Veliston', 'age': 21}


In [17]:
# Positional Arguments:

def greet(name, age):
    return f"Hello {name}, you are {age} years old."

print(greet("Veliston", 28))

# Keyword Arguments:

print(greet(age=20, name="Rocha")) 

Hello Veliston, you are 28 years old.
Hello Rocha, you are 20 years old.


In [18]:
#  Function Expressions

# functions can be assigned to variables

def add(x, y):
    return x + y

# assifning function to variable
operation = add

print(operation(4, 5))


9


### Nested Functions

> Functions defined inside other functions. They can access variables from the enclosing function.


In [19]:
def outer_function(x): # oter
    def inner_function(y): # inner
        return x + y
    return inner_function

# calling
func = outer_function(10)
print(func(5))

15


# 4. What is the purpose of the ' return ' statement in a python function?

> The '  return  ' statement is used in a function to send a value back to the caller of the function. 

In [23]:
# example:

def add_numbers(a, b):
    result = a + b
    return result #  with return statement

sum_value = add_numbers(5, 3)
print(sum_value)
print(type(sum_value))

8
<class 'int'>


In [26]:
# example
# without return value

def add_numbers(a, b):
    result = a + b
    print(result)

sum_value = add_numbers(5, 3)
print(sum_value)
print(type(sum_value))

# without return staement it return None

8
None
<class 'NoneType'>


# 5. What are iteration in Python and how do they differ from iterables?

### Iterable:

> An object that can return an iterator. Examples include lists, tuples, strings, dictionaries, and sets.

> Must implement __iter__() method which returns an iterator.

> Can be passed to functions like iter() to obtain an iterator.

### Iterator: 

> An object that represents a stream of data, it produces items one at a time and only once.

> Must implement both __iter__() and __next__() methods.

> Can be used in a loop (eg. for loop) to retrieve elements one at a time until the StopIteration exception is raised.

In [4]:
# Iterable: list
my_list = [1, 2, 3, 4]

# iterator
my_iterator = iter(my_list)

# Iterate through the iterator
print(next(my_iterator))
print(next(my_iterator))
print(next(my_iterator))
print(next(my_iterator))

# If we call next() again, it will raise StopIteration
try:
    print(next(my_iterator))
except StopIteration:
    print("End of iterator reached")

# Using  for loop to iterate over an iterable
for item in my_list:
    print(item)

1
2
3
4
End of iterator reached
1
2
3
4


# 6. Explain the concept of generators in Python and how they are defined.

>Generators are a way to create iterators.

>Instead of returning a single value and terminating like regular functions, generators yield values one at a time, pausing between each yield.

>They are defined using the yield keyword instead of return.

> They are memory efficient and can be used to iterate over large datasets.

In [6]:
# Generator function

def simple_generator():
    yield 1
    yield 2
    yield 3

gen = simple_generator()

# Iterating generator
print(next(gen))
print(next(gen)) 
print(next(gen))  



1
2
3


In [7]:
# Generator with loop

def simple_generator():
    yield 1
    yield 2
    yield 3

gen = simple_generator()

# Iterating generator using a loop
for value in gen:
    print(value)

1
2
3


In [9]:
# Generator expression

gen_expr = (x * x for x in range(3))

# Iterating  generator
print(next(gen_expr))
print(next(gen_expr))
print(next(gen_expr)) 

# it will cause error if  we run next(gen_expr) again

0
1
4


# 7. What are the advantages of using generators over regular functions?

>Memory Efficiency: Generators do not store the entire sequence in memory.

>Lazy Evaluation: Values are computed only when needed.

>Improved Performance: Reduced computation and faster startup times.

>State Retention: Generators maintain their state between iterations.

>Simpler Code: Easier to write and maintain complex iteration logic.

>Pipeline Construction: Generators can be composed into efficient data processing pipelines.

In [2]:
# using list
def squares_list(n):
    return [i**2 for i in range(n)]

squares = squares_list(17)
squares

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256]

In [31]:
# using generator
def squares_generator(n):
    for i in range(n):
        yield i**2

squares = squares_generator(10000)
squares

<generator object squares_generator at 0x000002262FD7C2B0>

# 8. What is lambda function in python and when is it typically used?

A lambda function in Python is a small anonymous function defined with the lambda keyword. Lambda functions can have any number of arguments but can only have one expression. They are syntactically restricted to a single expression and are typically used for simple operations that can be written in one line.

Syntax

>lambda arguments: expression

In [11]:
# example

add = lambda x, y: x + y
print(add(2, 3))

5


>Lambda functions are typically used for short, simple functions that are used temporarily, often passed as arguments to higher-order functions.

In [12]:
# Using filter() with a lambda function to filter a list:

numbers = [1, 2, 3, 4, 5, 6]
even_numbers = list(filter(lambda x: x % 2 == 0, numbers))
print(even_numbers)

[2, 4, 6]


In [13]:
# Using map() with a lambda function to transform a list:

numbers = [1, 2, 3, 4]
squared_numbers = list(map(lambda x: x ** 2, numbers))
print(squared_numbers)


[1, 4, 9, 16]


9. Explain the purpose and usage of the map(), reduce() and filter() functions in python?

### 1. map()

Purpose: map() applies a given function to all items in an input list (or any other iterable) and returns an iterator of the results.


Syntax
>map(function, iterable,)

>function: A function that takes one or more arguments.

>iterable: One or more iterables.

In [16]:
# Example: Squaring each number in a list

numbers = [1, 2, 3, 4]
squared_numbers = list(map(lambda x: x ** 2, numbers))
print(squared_numbers) 

numbers1 = [1, 2, 3]
numbers2 = [4, 5, 6]
sum_numbers = list(map(lambda x, y: x + y, numbers1, numbers2))
print(sum_numbers)  

[1, 4, 9, 16]
[5, 7, 9]


### 2. filter()

Purpose: filter() constructs an iterator from elements of an iterable for which a function returns true.

Syntax:

>filter(function, iterable)

>function: A function that takes one argument and returns a boolean.

>iterable: The iterable to be filtered.

In [17]:
# Example: Filtering even numbers from a list
numbers = [1, 2, 3, 4, 5, 6]
even_numbers = list(filter(lambda x: x % 2 == 0, numbers))
print(even_numbers)  

# Example: Filtering out non-positive numbers
numbers = [-1, 0, 1, 2, 3]
positive_numbers = list(filter(lambda x: x > 0, numbers))
print(positive_numbers)  


[2, 4, 6]
[1, 2, 3]


### 2. reduce()

Purpose: reduce() applies a rolling computation to sequential pairs of values in an iterable, effectively reducing the iterable to a single cumulative value. It is part of the functools module.

Syntax:

>functools.reduce(function, iterable[, initializer])

>function: A function that takes two arguments.

>iterable: The iterable to be reduced.

>initializer: (Optional) Initial value for the accumulator.



In [18]:
from functools import reduce

# Example: Calculating the product of all numbers in a list
numbers = [1, 2, 3, 4]
product = reduce(lambda x, y: x * y, numbers)
print(product)

# Example: Summing all numbers in a list 
total = reduce(lambda x, y: x + y, numbers, 10)
print(total)


24
20


10. What is the difference between map(), reduce() and filter() functions in python?

### map()

Key Points:

>Transforms each element in the iterable.

>The function is applied to each element individually.

>The resulting iterable has the same number of elements as the original.

In [19]:
numbers = [1, 2, 3, 4, 5, 6]
squared_numbers = list(map(lambda x: x ** 2, numbers))
print(squared_numbers) 

[1, 4, 9, 16, 25, 36]


### reduce()

Key Points:

>Combines elements of the iterable into a single value.

>The function takes two arguments: an accumulator and the next item in the sequence.

>Typically used for operations like summing, multiplying, or finding the greatest common divisor.

In [20]:
from functools import reduce
numbers = [1, 2, 3, 4, 5, 6]
product = reduce(lambda x, y: x * y, numbers)
print(product)

720


### filter()

Key Points:

>Selects elements based on a condition.

>The function is a predicate that returns True or False.

>The resulting iterable may have fewer elements than the original.

In [21]:
numbers = [1, 2, 3, 4, 5, 6]
even_numbers = list(filter(lambda x: x % 2 == 0, numbers))
print(even_numbers)

[2, 4, 6]


11. Using pen & paper write the internal mechanism for sum operating using reduce function on this given list: [47,11,42,13];

In [2]:
# Import the reduce function:

from functools import reduce


# Define the list:

numbers = [47, 11, 42, 13]


# Use reduce with a lambda function to sum the list:

result = reduce(lambda x, y: x + y, numbers)

result

113

#### Initial State:

* List: [47, 11, 42, 13]

* reduce starts by taking the first two elements of the list.


#### First Iteration:

* x = 47, y = 11
* x + y = 47 + 11 = 58
* Intermediate result: 58


#### Second Iteration:

* x = 58 (result of the first iteration), y = 42
* x + y = 58 + 42 = 100
* Intermediate result: 100


#### Third Iteration:

* x = 100 (result of the second iteration), y = 13
* x + y = 100 + 13 = 113
* Final result: 113

# Practical Questions

1. Write a Python function that takes a list of numbers as input and returns the sum of all even numbers in the list.

In [4]:
def sum_of_even_numbers(numbers):
  """Calculates the sum of even numbers in a list."""
  even_sum = 0
  for num in numbers:
    if num % 2 == 0:
      even_sum += num
  return even_sum

numbers = [1, 2, 3, 4, 5, 6]
print(sum_of_even_numbers(numbers))


12


2. Create a Python function that accepts a string and returns the reverse of that string.

In [9]:
def reverse_string(string):
  """Reverses a given string."""
  return string[::-1]

name = "Veliston Rocha"

reverse_string(name)

'ahcoR notsileV'

3. Implement a Python function that takes a list of integers and returns a new list containing the squares of each number.

In [10]:
def square_list(numbers):
  """Squares each number in a list and returns a new list."""
  squared_list = [num ** 2 for num in numbers]
  return squared_list

num = [1,3,5,7,9]
square_list(num)

[1, 9, 25, 49, 81]

4. Write a Python function that checks if a given number is prime or not from 1 to 200.

In [8]:
def is_prime(num):
  """Checks if a number is prime."""
  if num <= 1:
    return False
  if num <= 3:
    return True
  if num % 2 == 0 or num % 3 == 0:
    return False
  i = 5
  while i * i <= num:
    if num % i == 0 or num % (i + 2) == 0:
      return False
    i += 6
  return True

is_prime(17)


True

5. Create an iterator class in Python that generates the Fibonacci sequence up to a specified number of terms.

In [26]:
class FibonacciIterator:
    """Generates the Fibonacci sequence."""

    def __init__(self, num_terms):
        self.num_terms = num_terms 
        self.current_term = 0  
        self.next_term = 1 

    def __iter__(self):
        return self  

    def __next__(self):
        if self.num_terms <= 0:
            raise StopIteration  

        result = self.current_term  
        
        self.current_term, self.next_term = self.next_term, self.current_term + self.next_term
       
        self.num_terms -= 1
        
        return result


fib_seq = FibonacciIterator(10)  

for num in fib_seq:
    print(num)




0
1
1
2
3
5
8
13
21
34


6. Write a generator function in Python that yields the powers of 2 up to a given exponent.

In [24]:
def powers_of_two(exponent):
  """Generates powers of 2 up to a given exponent."""
  base = 2
  for i in range(exponent + 1):
    yield base ** i

expo = 10
print(f"Powers of 2 up to 2^{expo}:")
for value in powers_of_two(expo):
    print(value)

Powers of 2 up to 2^10:
1
2
4
8
16
32
64
128
256
512
1024


7. Implement a generator function that reads a file line by line and yields each line as a string.

In [23]:
def read_file(filepath):
    """Read file line by line."""
    try:
        with open(filepath, 'r') as file:
            for line in file:
                yield line.strip()
    except FileNotFoundError:
        print(f"The file at {filepath} was not found.")
    except Exception as e:
        print(f"An error occurred: {e}")

for line in read_file('example.txt'):
    print(line)

Python is a popular programming language.

Python can be used on a server to create web applications.


8. Use a lambda function in Python to sort a list of tuples based on the second element of each tuple.

In [1]:
tuples_list = [(1, 3), (4, 2), (2, 5)]
sorted_tuples = sorted(tuples_list, key=lambda x: x[1])
print(sorted_tuples)  


[(4, 2), (1, 3), (2, 5)]


9. Write a Python program that uses map() to convert a list of temperatures from Celsius to Fahrenheit.

In [37]:
def celsius_to_fahrenheit(celsius):
  """Converts Celsius to Fahrenheit."""
  return (celsius * 9/5) + 32

temperatures_celsius = [25, 30, 35]
temperatures_fahrenheit = list(map(celsius_to_fahrenheit, temperatures_celsius))
print(temperatures_fahrenheit)


[77.0, 86.0, 95.0]


10. Create a Python program that uses 'filter()' to remove all the vowels from a given string.

In [38]:
def remove_vowels(char):
  """Removes vowels from a character."""
  vowels = 'aeiouAEIOU'
  return char not in vowels

string = "Hello, World!"
without_vowels = ''.join(filter(remove_vowels, string))
print(without_vowels)


Hll, Wrld!


11. Accounting Routine for Book Shop:

In [18]:
orders = [
    ["34587", "Learning Python", "Mark Lutz", 4, 40.95],
    ["98762", "Programming Python", "Mark Lutz", 5, 56.80],
    ["77226", "Head First Python", "Paul Barry", 3, 32.95],
    ["88112", "Einfuhrung in Python3", "Bernd Klein", 3, 24.99]
]

order_summary = list(map(lambda x: (x[0], x[3] * x[4] + 10 if x[3] * x[4] < 100 else x[3] * x[4]), orders))

print(order_summary)

print("-" * 50)
print("-" * 50)
print("Order Summary:")
print("-" * 50)
print(f"{'Order Number':<15}{'Total Amount (€)':<15}")
print("-" * 50)

for order in order_summary:
    order_number, total_amount = order
    print(f"{order_number:<15}{total_amount:<15.2f}")

print("-" * 50)

[('34587', 163.8), ('98762', 284.0), ('77226', 108.85000000000001), ('88112', 84.97)]
--------------------------------------------------
--------------------------------------------------
Order Summary:
--------------------------------------------------
Order Number   Total Amount (€)
--------------------------------------------------
34587          163.80         
98762          284.00         
77226          108.85         
88112          84.97          
--------------------------------------------------
